File: perl/version3_intl.pl
# http://localhost:7072/?n=10
# perl version3_intl.pl

use strict;
use warnings;
use HTTP::Daemon;
use HTTP::Status;
use URI;

my $servidor = HTTP::Daemon->new(
    LocalAddr => 'localhost',
    LocalPort => 7072,
) or die "No se pudo iniciar el servidor: $!";

print "Servidor en http://localhost:7072/?n=10\n";

while (my $conexion = $servidor->accept()) {
    while (my $peticion = $conexion->get_request()) {
        my $uri = URI->new($peticion->uri);
        my %params = $uri->query_form;
        my $n = $params{n} // '0';

        my $resultado = convertir_espanol(int($n));

        my $respuesta = HTTP::Response->new(200);
        $respuesta->header('Content-Type' => 'text/plain; charset=utf-8');
        $respuesta->content($resultado);
        $conexion->send_response($respuesta);
    }
    $conexion->close();
}

sub convertir_espanol {
    my ($n) = @_;
    return 'cero' if $n == 0;
    return 'menos ' . convertir_espanol(-$n) if $n < 0;

    my @unidades = ('', 'uno', 'dos', 'tres', 'cuatro', 'cinco',
                    'seis', 'siete', 'ocho', 'nueve', 'diez', 'once', 'doce',
                    'trece', 'catorce', 'quince', "diecis\x{e9}is", 'diecisiete',
                    'dieciocho', 'diecinueve', 'veinte');
    my @decenas = ('', '', 'veinte', 'treinta', 'cuarenta',
                   'cincuenta', 'sesenta', 'setenta', 'ochenta', 'noventa');
    my @centenas = ('', 'ciento', 'doscientos', 'trescientos',
                    'cuatrocientos', 'quinientos', 'seiscientos', 'setecientos',
                    'ochocientos', 'novecientos');

    return $unidades[$n] if $n <= 20;
    return 'veinti' . $unidades[$n - 20] if $n < 30;

    if ($n < 100) {
        my $res = $decenas[int($n / 10)];
        $res .= ' y ' . $unidades[$n % 10] if $n % 10 != 0;
        return $res;
    }

    return 'cien' if $n == 100;

    if ($n < 1000) {
        my $res = $centenas[int($n / 100)];
        $res .= ' ' . convertir_espanol($n % 100) if $n % 100 != 0;
        return $res;
    }

    if ($n < 2000) {
        my $res = 'mil';
        $res .= ' ' . convertir_espanol($n % 1000) if $n % 1000 != 0;
        return $res;
    }

    if ($n < 1_000_000) {
        my $res = convertir_espanol(int($n / 1000)) . ' mil';
        $res .= ' ' . convertir_espanol($n % 1000) if $n % 1000 != 0;
        return $res;
    }

    return "$n";
}

File: cpp/version3_intl.cpp
// http://localhost:8102/?n=10
// cl version3_intl.cpp /link ws2_32.lib
// version3_intl.exe

#include <iostream>
#include <string>
#include <winsock2.h>

#pragma comment(lib, "ws2_32.lib")

std::string convertirEspanol(long long n) {
    if (n == 0) return "cero";
    if (n < 0) return "menos " + convertirEspanol(-n);

    std::string unidades[] = {"","uno","dos","tres","cuatro","cinco","seis",
        "siete","ocho","nueve","diez","once","doce","trece","catorce","quince",
        "diecis\xc3\xa9is","diecisiete","dieciocho","diecinueve","veinte"};
    std::string decenas[] = {"","","veinte","treinta","cuarenta","cincuenta",
        "sesenta","setenta","ochenta","noventa"};
    std::string centenas[] = {"","ciento","doscientos","trescientos","cuatrocientos",
        "quinientos","seiscientos","setecientos","ochocientos","novecientos"};

    if (n <= 20) return unidades[n];
    if (n < 30) return "veinti" + unidades[n-20];
    if (n < 100) {
        std::string r = decenas[n/10];
        if (n%10) r += " y " + unidades[n%10];
        return r;
    }
    if (n == 100) return "cien";
    if (n < 1000) {
        std::string r = centenas[n/100];
        if (n%100) r += " " + convertirEspanol(n%100);
        return r;
    }
    if (n < 2000) {
        std::string r = "mil";
        if (n%1000) r += " " + convertirEspanol(n%1000);
        return r;
    }
    if (n < 1000000) {
        std::string r = convertirEspanol(n/1000) + " mil";
        if (n%1000) r += " " + convertirEspanol(n%1000);
        return r;
    }
    return std::to_string(n);
}

void manejarCliente(SOCKET cliente) {
    char buffer[4096] = {0};
    recv(cliente, buffer, 4096, 0);
    std::string peticion(buffer);

    std::string n = "0";
    size_t pos = peticion.find("n=");
    if (pos != std::string::npos) {
        size_t fin = peticion.find(" ", pos);
        n = peticion.substr(pos + 2, fin - pos - 2);
    }

    std::string resultado = convertirEspanol(std::stoll(n));
    std::string respuesta = "HTTP/1.1 200 OK\r\nContent-Type: text/plain; charset=utf-8\r\n\r\n" + resultado;
    send(cliente, respuesta.c_str(), respuesta.size(), 0);
    closesocket(cliente);
}

int main() {
    WSADATA wsa;
    WSAStartup(MAKEWORD(2, 2), &wsa);
    SOCKET servidor = socket(AF_INET, SOCK_STREAM, 0);
    sockaddr_in addr;
    addr.sin_family = AF_INET;
    addr.sin_port = htons(8102);
    addr.sin_addr.s_addr = INADDR_ANY;
    bind(servidor, (sockaddr*)&addr, sizeof(addr));
    listen(servidor, 5);
    std::cout << "Servidor en http://localhost:8102/?n=10" << std::endl;
    while (true) {
        SOCKET cliente = accept(servidor, nullptr, nullptr);
        manejarCliente(cliente);
    }
    WSACleanup();
    return 0;
}